12-offer-encoding.md

```diff
@@ -0,0 +1,870 @@
+   to prove an invoice without revealing its entirety.
+3. Fields cannot generally be extracted for external use: the `h`
+   field was a boutique extraction of the `d` field only.
+4. The lack of the 'it's OK to be odd' rule makes backward compatibility
+   harder.
+5. The 'human-readable' idea of separating amounts proved fraught:
+   `p` was often mishandled, and amounts in pico-bitcoin are harder
+   than the modern satoshi-based counting.
+6. Developers found the bech32 encoding to have an issue with extensions,
+   which means we want to replace or discard it anyway.
+7. The `payment_secret` designed to prevent probing by other nodes in
+   the path was only useful if the invoice remained private between the
+   payer and payee.
+8. Invoices must be given per user and are actively dangerous if two
+   payment attempts are made for the same user.
+
+
+# Payment Flow Scenarios
+
+Here we use "user" as shorthand for the individual user's lightning
+node and "merchant" as the shorthand for the node of someone who is
+selling or has sold something.
+
+There are two basic payment flows supported by BOLT 12:
+
+The general user-pays-merchant flow is:
+1. A merchant publishes an *offer*, such as on a web page or a QR code.
+2. Every user requests a unique *invoice* over the lightning network
+   using an *invoice_request* message, which contains the offer fields.
+3. The merchant replies with the *invoice*.
+4. The user makes a payment to the merchant as indicated by the invoice.
+
+The merchant-pays-user flow (e.g. ATM or refund):
+1. The merchant publishes an *invoice_request* which contains offer fields
+   which refer to its attempt to send money, including an amount.
+2. The user sends an *invoice* over the lightning network for the amount in the
+   *invoice_request*, using a (possibly temporary) *invoice_node_id*.
+3. The merchant confirms the *invoice_node_id* to ensure it's about to pay the correct
+   person, and makes a payment to the invoice.
+
+## Payment Proofs and Payer Proofs
+
+Note that the normal lightning "proof of payment" can only demonstrate that an
+invoice was paid (by showing the preimage of the `payment_hash`), not who paid
+it.  The merchant can claim an invoice was paid, and once revealed, anyone can
+claim they paid the invoice, too.[1]
+
+Providing a key in *invoice_request* allows the payer to prove that they were the one
+to request the invoice.  In addition, the Merkle construction of the BOLT 12
+invoice signature allows the user to reveal invoice fields in case
+of a dispute selectively.
+
+# Encoding
+
+Each of the forms documented here are in
+[TLV](01-messaging.md#type-length-value-format) format.
+
+The supported ASCII encoding is the human-readable prefix, followed by a
+`1`, followed by a bech32-style data string of the TLVs in order,
+optionally interspersed with `+` (for indicating additional data is to
+come).  There is no checksum, unlike bech32m.
+
+## Requirements
```

Comment:
Hi, I think it might be a good idea to explicitly specify here that QR Codes should be encoded in UPPERCASE to save space and that lowercase should be used for textual presentation, similar to the bolt 11 spec.

I bring this up because some implementations have had issues recently.  cln-application was unable to parse the UPPERCASE invoices or offers that were generated by specifically the iOS version of Phoenix wallet, which was copying the presentation text in UPPERCASE.  And there are questions on the Phoenix side (https://github.com/ACINQ/phoenix/pull/610#pullrequestreview-2234129467) whether bolt12 QR Codes should be rendered in UPPERCASE.  I think the answer is yes?  If so can we add this to the formal spec to be clear to the wallet implementations?

> https://github.com/lightning/bolts/blob/master/11-payment-encoding.md
```
A writer:

MUST encode the payment request in Bech32 (see BIP-0173)
SHOULD use upper case for QR codes (see BIP-0173)
MAY exceed the 90-character limit specified in BIP-0173.
```
> https://en.bitcoin.it/wiki/BIP_0173 
```
For presentation, lowercase is usually preferable, but inside QR codes uppercase SHOULD be used, as those permit the use of [alphanumeric mode](http://www.thonky.com/qr-code-tutorial/alphanumeric-mode-encoding), which is 45% more compact than the normal [byte mode](http://www.thonky.com/qr-code-tutorial/byte-mode-encoding).
```